device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsIotHubConnection.java

```diff
@@ -468,9 +470,29 @@ public void onAuthenticationSessionOpened()
     }
 
     @Override
-    public void onMessageAcknowledged(Message message)
+    public void onMessageAcknowledged(Message message, DeliveryState deliveryState)
     {
-        this.listener.onMessageSent(message, null);
+        if (deliveryState == Accepted.getInstance())
+        {
+            this.listener.onMessageSent(message, null);
+        }
+        else if (deliveryState instanceof Rejected)
+        {
+            // The message was not accepted by the server, and the reason why is found within the nested error
+            TransportException ex = AmqpsExceptionTranslator.convertFromAmqpException(((Rejected) deliveryState).getError());
+            this.listener.onMessageSent(message, ex);
+        }
+        else if (deliveryState == Released.getInstance())
```

Comment:
Does the service ever send a delivery state of "released"? What do they expect the client to do, in this case?

Reply:
I've never seen it, but the protocol defines it as a possible delivery state. As a delivery state, it just means that the service has made the message available to be re-delivered for whatever reason. So on the client side, we just need to create a retryable exception here so that the transport layer re-sends it